clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java

```diff
@@ -437,15 +442,21 @@ public boolean isTransactional() {
         return transactionalId != null;
     }
 
-    // Check all the finalized features from apiVersions to whether the transaction V2 is enabled.
+    /**
+     *  Check all the finalized features from apiVersions to whether the transaction V2 is enabled.
+     *  Sets isUpgradingToV2 if the previous value was false and now it is true.
+     */
+
     public synchronized void maybeUpdateTransactionV2Enabled() {
         if (latestFinalizedFeaturesEpoch >= apiVersions.getMaxFinalizedFeaturesEpoch()) {
             return;
         }
         ApiVersions.FinalizedFeaturesInfo info = apiVersions.getFinalizedFeaturesInfo();
         latestFinalizedFeaturesEpoch = info.finalizedFeaturesEpoch;
         Short transactionVersion = info.finalizedFeatures.get("transaction.version");
+        boolean previousValue = isTransactionV2Enabled;
         isTransactionV2Enabled = transactionVersion != null && transactionVersion >= 2;
+        isUpgradingToV2 = currentState != State.READY && !previousValue && isTransactionV2Enabled;
```

Comment:
What is the reason for this condition? I may miss some cases but maybeUpdateTransactionV2Enabled is only called in initialization and commit transaction, none of which set the state to be Ready? When we will have `currentState == State.READY`

Reply:
On the first initialization we actually "upgrade to v2" because the default is false 😅. We don't want the logic there though.

Reply:
That makes sense 👍 will change that.